source/qnx/ionotify_event_loop.c

```diff
@@ -0,0 +1,1042 @@
+
+static bool s_is_on_callers_thread(struct aws_event_loop *event_loop) {
+    struct ionotify_loop *ionotify_loop = event_loop->impl_data;
+
+    aws_thread_id_t *thread_id = aws_atomic_load_ptr(&ionotify_loop->running_thread_id);
+    return thread_id && aws_thread_thread_id_equal(*thread_id, aws_thread_current_thread_id());
+}
+
+static void s_process_task_pre_queue(struct aws_event_loop *event_loop) {
+    struct ionotify_loop *ionotify_loop = event_loop->impl_data;
+
+    AWS_LOGF_TRACE(AWS_LS_IO_EVENT_LOOP, "id=%p: Processing cross-thread tasks", (void *)event_loop);
+
+    struct aws_linked_list task_pre_queue;
+    aws_linked_list_init(&task_pre_queue);
+
+    aws_mutex_lock(&ionotify_loop->task_pre_queue_mutex);
+    aws_linked_list_swap_contents(&ionotify_loop->task_pre_queue, &task_pre_queue);
+    aws_mutex_unlock(&ionotify_loop->task_pre_queue_mutex);
+
+    while (!aws_linked_list_empty(&task_pre_queue)) {
+        struct aws_linked_list_node *node = aws_linked_list_pop_front(&task_pre_queue);
+        struct aws_task *task = AWS_CONTAINER_OF(node, struct aws_task, node);
+        AWS_LOGF_TRACE(
+            AWS_LS_IO_EVENT_LOOP,
+            "id=%p: task %p pulled to event-loop, scheduling now.",
+            (void *)event_loop,
+            (void *)task);
+        /* Timestamp 0 is used to denote "now" tasks */
+        if (task->timestamp == 0) {
+            aws_task_scheduler_schedule_now(&ionotify_loop->scheduler, task);
+        } else {
+            aws_task_scheduler_schedule_future(&ionotify_loop->scheduler, task, task->timestamp);
+        }
+    }
+}
+
+/**
+ * This just calls MsgReceive().
+ *
+ * We broke this out into its own function so that the stacktrace clearly shows
+ * what this thread is doing. We've had a lot of cases where users think this
+ * thread is deadlocked because it's stuck here. We want it to be clear
+ * that it's doing nothing on purpose. It's waiting for events to happen...
+ */
+AWS_NO_INLINE
+static rcvid_t aws_event_loop_listen_for_io_events(
+    int io_events_channel_id,
+    const uint64_t *timeout,
+    struct _pulse *pulse,
+    int *errno_value) {
+    /* Event of type SIGEV_UNBLOCK makes the timed-out kernel call fail with an error of ETIMEDOUT. */
+    struct sigevent notify;
+    SIGEV_UNBLOCK_INIT(&notify);
+    int rc = TimerTimeout(CLOCK_MONOTONIC, _NTO_TIMEOUT_RECEIVE, &notify, timeout, NULL);
+    if (rc == -1) {
+        *errno_value = errno;
+        return rc;
+    }
+    rcvid_t rcvid = MsgReceive(io_events_channel_id, pulse, sizeof(*pulse), NULL);
+    if (rcvid == -1) {
+        *errno_value = errno;
+    }
+    return rcvid;
+}
+
+static void s_aws_ionotify_cleanup_aws_lc_thread_local_state(void *user_data) {
+    (void)user_data;
+    aws_cal_thread_clean_up();
+}
+
+static void s_process_pulse(
+    struct aws_event_loop *event_loop,
+    const struct _pulse *pulse,
+    bool *should_process_cross_thread_tasks) {
+    if (pulse->code == CROSS_THREAD_PULSE_SIGEV_CODE) {
```

Comment:
This feels like maybe it should be up in the calling code instead, removing the awkward output param.